### ISSUE TITLE
Forecasting at scale (in parallel)

### ISSUE THREAD

**Issue by sabersf** [facebook/prophet#479]
I'd like to use prophet to predict time series for different entities over the same time frame (these entities are not related so it's not a multivariate problem). i.e. let's say I have 1,000 features that change over the same period of time, and I'd like to predict their values for the next 30 days. How can I train these models in parallel (or distribute the computation) using Prophet? Is there any example that covers something similar?

**Comment by leafonsword** [comment]
I found `forecast = m.predict(future)` this statement is slow in python when dataframe contains many rows, could add a parallel options so all cpu cores could compute parallelly~

**Comment by bletham** [comment]
The `predict` can be slow if `future` has a lot of rows because of the simulation for trend forecast uncertainty. You can speed things up by using fewer samples to estimate trend uncertainty intervals - the argument `uncertainty_samples` to `Prophet()` defaults to 1000, and cutting that down will make it run faster at a cost of more noise in the upper and lower bounds.

Otherwise, if the time series are all totally independent then you could always just run a process for each processor separately on a subset of time series.

**Comment by bletham** [comment]
I think that `predict_uncertainty` is the only one that should take substantial resources, but anyway agree that it'd be good to speed these up.

We've been talking recently about the possibility of moving the prediction code into Stan, which would speed it up a bunch and also reduce some code duplication between R and Python versions. I think that'd be the nicest way to improve prediction speed.

**Comment by leafonsword** [comment]
@bletham 
prediction code into Stan centainly will have a huge speedup!

I make a `predict()` which takes about 5s, and these 3 sub-function takes about:
```
 predict_trend() # less than 1 s
predict_seasonal_components() # about 2s
predict_uncertainty()   # about 3s
```

Since `predict_seasonal_components()` takes about 40% time, parallel execuing will sppedup about 40% comparing with in order executing whatever prediction code into Stan or not~

**Comment by bletham** [comment]
That's pretty beneficial. I'd be a bit hesitant to lose Py2 compatibility though, at least around here the transition is still a work in progress. I think I'd like to first evaluate the feasibility of doing the predictions in Stan.

**Comment by leafonsword** [comment]
@bletham 
python2 also has a concurrent.futures lib：
https://github.com/agronholm/pythonfutures

Anyway，predictions in Stan is first priority~

**Comment by bletham** [comment]
It looks like predictions in Stan are going to possible once an upstream issue in RStan has been fixed (#501). I'm going to close this issue since we are definitely moving in that direction, and once that is done we can revisit parallelization.
